Major refactoring of code.

bsutton opened the pull request:
Added an assert when an attempt is made to transition to a non-registered state. Previously it was throwing a noSuchMethod exception.

bsutton committed:
Added an assert when an attempt is made to tranistion to a non-registered state. Previously it was throwing a noSuchMethod exception.

bsutton committed:
reverted formatting to 80 chars.

ookami-kb reviewed (changes requested):
Thanks for the PR. I actually think it would be better to replace it with throwing an error, as assert won't work in release.

Smth like this:

```dart
if (!_graph.stateDefinitions.containsKey(v.toState.runtimeType)) {
  throw StateError('State ${v.toState} is not registered');
}
```

bsutton commented:
O was rooting with the idea of an exception but didn't want to change the API.
I will revise and push a new or.

bsutton committed:
When a state is encountered that is unregistered we now throw an UnregisteredState rather than simply asserting.

bsutton committed:
Formatted code back to 80 chars.

bsutton commented:
I've used a unique exception rather than reusing a generic exception. Its generally recommend to uses specific exception as it makes it easier for a user of the api to catch and process.

bsutton committed:
removed dfunc library as it was just complicating things.

bsutton committed:
Major refactoring of code.
Introduced condtional event transitions.
Changed SideEffects to be a callback rather than an object.
Renamed a number of methods and classes for greater clarity.
Introduced State and Event interfaces to simplify the Generics.

bsutton committed:
released version 0.1.0

bsutton commented:
This is now a major refactoring of the api.

I will completely understand if you don't want to except this PR given the scale of the changes and the rather opinionated approach I've taken to the changes :)

If you find these too much let me know and I will released the code from my own fork.

I've not I've removed the builders as I there were making maintenance more complicated and didn't really add any value to the core fsm project.

The unit test all pass.

bsutton committed:
Fixed a bug where the 'State' passed to the EventHandler was 'State' rather than the derived type.

bsutton committed:
Fixed a bug where the 'State' passed to the EventHandler was 'State' rather than the derived type.

bsutton committed:
Merge branch 'master' of github.com:bsutton/fsm

bsutton committed:
changed from extends to implements for State and Events to serve as a better example.

ookami-kb commented:
Yeah, I don't think that this is the way it should be done and it pretty much changes the concept of the library. So it really makes sense to publish it as a separate library.
